ako-gateway-api/k8s/gateway_controller.go

```diff
@@ -103,6 +110,86 @@ func (c *GatewayController) SetupEventHandlers(k8sinfo k8s.K8sinformers) {
 	c.workqueue = mcpQueue.Workqueue
 	numWorkers := mcpQueue.NumWorkers
 
+	// Add EPSInformer
+	epsEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			eps := obj.(*discovery.EndpointSlice)
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(eps))
+			svcName, ok := eps.Labels[discovery.LabelServiceName]
+			if !ok || svcName == "" {
+				utils.AviLog.Debugf("Endpointslice Add event: Endpointslice does not have backing svc")
+				return
+			}
+			key := utils.Endpointslices + "/" + namespace + "/" + svcName
+			if lib.IsNamespaceBlocked(namespace) {
+				utils.AviLog.Debugf("key: %s, msg: Endpoint Add event: Namespace: %s didn't qualify filter", key, namespace)
+				return
+			}
+			bkt := utils.Bkt(namespace, numWorkers)
+			c.workqueue[bkt].AddRateLimited(key)
+			utils.AviLog.Debugf("key: %s, msg: ADD", key)
+		},
+		DeleteFunc: func(obj interface{}) {
+			if c.DisableSync {
+				return
+			}
+			eps, ok := obj.(*discovery.EndpointSlice)
+			if !ok {
+				// endpoints were deleted but its final state is unrecorded.
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					utils.AviLog.Errorf("couldn't get object from tombstone %#v", obj)
+					return
+				}
+				eps, ok = tombstone.Obj.(*discovery.EndpointSlice)
+				if !ok {
+					utils.AviLog.Errorf("Tombstone contained object that is not an Endpointslice: %#v", obj)
+					return
+				}
+			}
+			namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(eps))
+			svcName, ok := eps.Labels[discovery.LabelServiceName]
+			if !ok || svcName == "" {
+				utils.AviLog.Debugf("Endpointslice Delete event: Endpointslice does not have backing svc")
+				return
+			}
+			key := utils.Endpointslices + "/" + namespace + "/" + svcName
+			if lib.IsNamespaceBlocked(namespace) {
+				utils.AviLog.Debugf("key: %s, msg: Endpointslice Delete event: Namespace: %s didn't qualify filter", key, namespace)
+				return
+			}
+			bkt := utils.Bkt(namespace, numWorkers)
+			c.workqueue[bkt].AddRateLimited(key)
+			utils.AviLog.Debugf("key: %s, msg: DELETE", key)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			if c.DisableSync {
+				return
+			}
+			oeps := old.(*discovery.EndpointSlice)
+			ceps := cur.(*discovery.EndpointSlice)
+			if oeps.ResourceVersion != ceps.ResourceVersion {
+				namespace, _, _ := cache.SplitMetaNamespaceKey(utils.ObjKey(ceps))
+				svcName, ok := ceps.Labels[discovery.LabelServiceName]
+				if !ok || svcName == "" {
+					utils.AviLog.Debugf("Endpointslice Delete event: Endpointslice does not have backing svc")
+					return
+				}
```

Comment:
this seems to not acknowledge transition of ep slice from non empty svcName to empty svcName. AKO will assume the svc is in old state. even when the ep slice gets deleted the key won't get pushed because of same validation in deleteFunc.

we should change this if not intended

Reply:
i have updated the code to check for svcName from old endpoint slice object in case there is a transition. For delete func if there is no svc name, there is no way to identify the endpoint slice. Its a dangling resource, even kubernetes does nothing if the svcName label is removed from endpointslice.